pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,471 @@
+func (e *EventedPLEG) Start() {
+	e.runningMu.Lock()
+	defer e.runningMu.Unlock()
+	if !IsEventedPLEGInUse {
+		IsEventedPLEGInUse = true
+		e.stopCh = make(chan struct{})
+		go wait.Until(e.watchEventsChannel, 0, e.stopCh)
+	}
+}
+
+func (e *EventedPLEG) Stop() {
+	e.runningMu.Lock()
+	defer e.runningMu.Unlock()
+	if IsEventedPLEGInUse {
+		close(e.stopCh)
+		IsEventedPLEGInUse = false
+	}
+}
+
+func (e *EventedPLEG) Update(relistingPeriod time.Duration, relistThreshold time.Duration) {
+	e.genericPleg.Update(relistingPeriod, relistThreshold)
+}
+
+// Healthy check if PLEG work properly.
+func (e *EventedPLEG) Healthy() (bool, error) {
+	// GenericPLEG is declared unhealthy when relisting time is more than
+	// the relistThreshold. In case, EventedPLEG is turned on the value of
+	// relistingPeriod, relistingThreshold is adjusted to higher values and
+	// so health of Generic PLEG should take care of the condition checking
+	// higher relistThreshold between last two relist.
+
+	// EventedPLEG is declared unhealthy only if eventChannel is out of capacity.
+	if len(e.eventChannel) >= cap(e.eventChannel) {
+		return false, fmt.Errorf("EventedPLEG: pleg event channel capacity is full with %v events", len(e.eventChannel))
+	}
+
+	timestamp := e.clock.Now()
+	metrics.PLEGLastSeen.Set(float64(timestamp.Unix()))
+	return true, nil
+}
+
+func (e *EventedPLEG) watchEventsChannel() {
+
+	containerEventsResponseCh := make(chan *runtimeapi.ContainerEventResponse, cap(e.eventChannel))
+	defer close(containerEventsResponseCh)
+
+	// Get the container events from the runtime.
+	go func() {
+		numAttempts := 0
+		for {
+			err := e.runtimeService.GetContainerEvents(containerEventsResponseCh)
+			if err == io.EOF {
+				// The runtime service has been closed, exit the goroutine.
+				klog.V(4).Info("Evented PLEG: Runtime service has been closed, exit the goroutine")
+				return
+			}
+
+			if err != nil && numAttempts <= e.eventedPlegMaxStreamRetries {
+				numAttempts++
+				e.Relist() // Force a relist to get the latest container and pods running metric.
+				klog.V(4).Infof("Evented PLEG: Failed to get container events, retrying: %v", err)
+				continue
+			}
+
+			if numAttempts >= e.eventedPlegMaxStreamRetries {
+				if IsEventedPLEGInUse {
+					klog.ErrorS(err, "Evented PLEG: Failed to get container events")
+					IsEventedPLEGInUse = false
+					// Fall back to Generic PLEG relisting since Evented PLEG is not working.
+					e.genericPleg.Stop()                                              // Stop the existing Generic PLEG which runs with longer relisting period when Evented PLEG is in use.
+					e.Stop()                                                          // Stop the evented PLEG too
+					e.Update(e.genericPlegRelistPeriod, e.genericPlegRelistThreshold) // Update the relisting period to the default value for the Generic PLEG.
+					e.genericPleg.Start()
+					break
+				}
+			}
+		}
+	}()
+
+	e.processCRIEvents(containerEventsResponseCh)
```

Comment:
If e.stop calls, I think this line should exit again

Reply:
agree